x-pack/plugins/security/server/lib/authentication/providers/kerberos.ts

```diff
@@ -0,0 +1,280 @@
+      }
+    }
+
+    // If we couldn't authenticate by means of all methods above, let's try to check if Elasticsearch can
+    // start authentication mechanism negotiation, otherwise just return authentication result we have.
+    return authenticationResult.notHandled()
+      ? await this.authenticateViaSPNEGO(request, state)
+      : authenticationResult;
+  }
+
+  /**
+   * Invalidates access token retrieved in exchange for SPNEGO token if it exists.
+   * @param request Request instance.
+   * @param state State value previously stored by the provider.
+   */
+  public async deauthenticate(request: Legacy.Request, state?: ProviderState) {
+    this.debug(`Trying to deauthenticate user via ${request.url.path}.`);
+
+    if (!state || !state.accessToken) {
+      this.debug('There is no access token invalidate.');
+      return DeauthenticationResult.notHandled();
+    }
+
+    try {
+      const {
+        invalidated_tokens: invalidatedAccessTokensCount,
+      } = await this.options.client.callWithInternalUser('shield.deleteAccessToken', {
+        body: { token: state.accessToken },
+      });
+
+      if (invalidatedAccessTokensCount === 0) {
+        this.debug('User access token was already invalidated.');
+      } else if (invalidatedAccessTokensCount === 1) {
+        this.debug('User access token has been successfully invalidated.');
+      } else {
+        this.debug(
+          `${invalidatedAccessTokensCount} user access tokens were invalidated, this is unexpected.`
+        );
+      }
+    } catch (err) {
+      this.debug(`Failed invalidating user's access token: ${err.message}`);
+      return DeauthenticationResult.failed(err);
+    }
+
+    return DeauthenticationResult.redirectTo('/logged_out');
+  }
+
+  /**
+   * Validates whether request contains `Negotiate ***` Authorization header and just passes it
+   * forward to Elasticsearch backend.
+   * @param request Request instance.
+   */
+  private async authenticateViaHeader(request: Legacy.Request) {
+    this.debug('Trying to authenticate via header.');
+
+    const authorization = request.headers.authorization;
+    if (!authorization) {
+      this.debug('Authorization header is not presented.');
+      return AuthenticationResult.notHandled();
+    }
+
+    // First attempt to exchange SPNEGO token for an access token.
+    let accessToken: string;
+    try {
+      accessToken = (await this.options.client.callWithRequest(request, 'shield.getAccessToken', {
+        body: { grant_type: 'client_credentials' },
+      })).access_token;
+    } catch (err) {
+      this.debug(`Failed to exchange SPNEGO token for an access token: ${err.message}`);
+      return AuthenticationResult.failed(err);
+    }
+
+    this.debug('Get token API request to Elasticsearch successful');
+
+    // We validate that access token exists in the response so other private methods in this class
+    // can rely on them both existing.
+    if (!accessToken) {
```

Comment:
__note:__ I'm copying this code from token provider, but probably it doesn't make sense... What do you think?

Reply:
It doesn't make sense because we really should always get an access token?

Reply:
Yeah, I believe so. Will remove this check